atom/browser/api/atom_api_web_contents.cc

```diff
@@ -884,26 +901,80 @@ void WebContents::DidGetRedirectForResourceRequest(
        details.headers.get());
 }
 
+void WebContents::DidStartNavigation(
+    content::NavigationHandle* navigation_handle) {
+  int frame_tree_node_id = navigation_handle->GetFrameTreeNodeId();
+  content::FrameTreeNode* frame_tree_node =
+    content::FrameTreeNode::GloballyFindByID(frame_tree_node_id);
+  content::RenderFrameHostManager* render_manager
+    = frame_tree_node->render_manager();
+  content::RenderFrameHostImpl* current_frame_host
+    = render_manager->current_frame_host();
+  content::RenderFrameHostImpl* pending_frame_host
+    = render_manager->pending_frame_host();
+  content::RenderFrameHostImpl* speculative_frame_host
```

Comment:
You can cast these to `RenderFrameHost`, its best to not rely on `*Impl` classes directly, there are some use cases of `WebContentsImpl` in the code base but they are unavoidable. This can simply be

```
+++ b/atom/browser/api/atom_api_web_contents.cc
@@ -56,9 +56,7 @@
 #include "chrome/browser/printing/print_view_manager_basic.h"
 #include "chrome/browser/ssl/security_state_tab_helper.h"
 #include "content/browser/frame_host/frame_tree_node.h"
-#include "content/browser/frame_host/render_frame_host_impl.h"
 #include "content/browser/frame_host/render_frame_host_manager.h"
-#include "content/browser/loader/global_routing_id.h"
 #include "content/browser/renderer_host/render_widget_host_impl.h"
 #include "content/browser/renderer_host/render_widget_host_view_base.h"
 #include "content/common/view_messages.h"
@@ -908,22 +906,18 @@ void WebContents::DidStartNavigation(
     content::FrameTreeNode::GloballyFindByID(frame_tree_node_id);
   content::RenderFrameHostManager* render_manager
     = frame_tree_node->render_manager();
-  content::RenderFrameHostImpl* current_frame_host
+  content::RenderFrameHost* current_frame_host
     = render_manager->current_frame_host();
-  content::RenderFrameHostImpl* pending_frame_host
-    = render_manager->pending_frame_host();
-  content::RenderFrameHostImpl* speculative_frame_host
+  content::RenderFrameHost* speculative_frame_host
     = render_manager->speculative_frame_host();
-  content::RenderFrameHostImpl* frame_host =
+  content::RenderFrameHost* frame_host =
     speculative_frame_host ? speculative_frame_host : current_frame_host;
 
   bool is_main_frame = navigation_handle->IsInMainFrame();
   int frame_process_id = -1, frame_routing_id = -1;
   if (frame_host) {
-    content::GlobalFrameRoutingId global_routing_id =
-      frame_host->GetGlobalFrameRoutingId();
-    frame_process_id = global_routing_id.child_id;
-    frame_routing_id = global_routing_id.frame_routing_id;
+    frame_process_id = frame_host->GetProcess()->GetID();
+    frame_routing_id = frame_host->GetRoutingID();
   }
   bool is_same_document = navigation_handle->IsSameDocument();
   auto url = navigation_handle->GetURL();
```

Reply:
sorry didn't notice your example